Assign cohorts to groups in the instructor dashboard

andy-armstrong opened the pull request:
TNL-653

andy-armstrong commented on openedx/core/djangoapps/course_groups/views.py:
```diff
@@ -29,76 +31,119 @@ def json_http_response(data):
+                _get_cohort_representation(c, course)
+                for c in cohorts.get_course_cohorts(course)
+            ]
+            # TODO: change to just directly returning the lists.
```
I dug around and couldn't find any consistency. The API documentation suggests that we should consider making this a paginated interface, but doesn't say what that looks like:

https://openedx.atlassian.net/wiki/display/AC/API+Thoughts

I think you should probably leave it as is for now, and in the future we should start to use the Django REST Framework for all of our APIs. For an example, see how DedX is using it in the user_api:

https://github.com/edx/edx-platform/blob/master/openedx/core/djangoapps/user_api/views.py#L168

andy-armstrong commented on openedx/core/djangoapps/course_groups/views.py:
```diff
@@ -29,76 +31,119 @@ def json_http_response(data):
+        else:
+            cohort = cohorts.get_cohort_by_id(course_key, cohort_id)
+            return JsonResponse(_get_cohort_representation(cohort, course))
+    else:
```
Makes sense. It is still odd that we are accepting posts, puts and patches and treating them all the same way. We should get consistent about this, which we will be forced to do if we use the Django REST Framework since you implement the POST methods directly. For example:

https://github.com/edx/edx-platform/blob/master/openedx/core/djangoapps/user_api/views.py#L253

andy-armstrong commented on test_root/uploads/nonunique_store.txt:
```diff
@@ -0,0 +1 @@
+copy
```
It looks like @clrux checked this in accidentally. I've removed them.

andy-armstrong commented on lms/static/js/spec/groups/views/cohorts_spec.js:
```diff
@@ -200,75 +266,60 @@ define(['backbone', 'jquery', 'js/common_helpers/ajax_helpers', 'js/common_helpe
-                    cohortsView.$('.action-save').click();
+                    getAddModal().find('.action-save').click();
                     expect(requests.length).toBe(0);
                     verifyMessage('Please enter a name for your new cohort group.', 'error');
```
@catong Thanks for the detailed explanation. I'm fine with doing this incrementally as you suggest.  Is this in our own style guide so that other teams also stop using "please"?

dan-f commented on lms/static/js/groups/views/cohorts.js:
```diff
@@ -45,13 +53,17 @@
                 cohortNavElement = this.$('.cohort-management-nav'),
-                additionalCohortControlElement = this.$('.wrapper-cohort-supplemental');
+                additionalCohortControlElement = this.$('.wrapper-cohort-supplemental'),
+                isModelUpdate = options && options.patch && response.hasOwnProperty('user_partition_id');
```
You could possibly override the sync method of the cohort model (or collection) to set a property indicating whether it was a model or collection sync.

dan-f commented on lms/static/js/groups/views/cohorts.js:
```diff
@@ -115,21 +129,32 @@
             this.removeNotification();
-            this.addCohortForm = $(this.addCohortFormTemplate({}));
-            this.addCohortForm.insertAfter(this.$('.cohort-management-nav'));
+            newCohort = new CohortModel();
```
Why can't the `cohortFormView` share the `cohortView`'s model?

dan-f commented on openedx/core/djangoapps/course_groups/views.py:
```diff
@@ -29,76 +31,119 @@ def json_http_response(data):
 
 def split_by_comma_and_whitespace(cstr):
     """
-    Split a string both by commas and whitespice.  Returns a list.
```
Hah!

dan-f commented on openedx/core/djangoapps/course_groups/views.py:
```diff
@@ -29,76 +31,119 @@ def json_http_response(data):
+        If no cohort ID is specified, returns the JSON representation of all cohorts.
+           This is returned as a dict with the list of cohort information stored under the
+           key `cohorts`.
+    PUT or POST or PATCH
```
I'm no REST expert, but I think POST makes sense in the case where a client POSTs to `courses/<course_key>/cohorts`. However, when creating a new cohort at `courses/<course_key>/cohorts/<new_cohort_id>`, PUT should be used.

Source: http://stackoverflow.com/questions/630453/put-vs-post-in-rest

andy-armstrong commented:
@explorerleslie @talbs @clrux What do you think of @catong's suggestion to switch to a check box instead of two radio buttons? See above ^^^

clrux commented:
@andy-armstrong How are we doing with this? Do you need anything from me?

explorerleslie commented:
@catong @andy-armstrong I'm not a fan of having a single checkbox that toggles on and off a feature. I'd prefer to keep the radio buttons. Maybe we could have something like: 

(field label) Content Group
( ) No Content Group
( ) Select a content group (dropdown)

cahrens commented:
@clrux did you see my notes above (https://github.com/edx/edx-platform/pull/6308#issuecomment-68449113)? I haven't checked out the branch lately, so perhaps some of those have already been addressed (or voted down). I was concerned though about the very small text size for the warning message, and wondered if it would have accessibility issues.

catong commented:
@explorerleslie @andy-armstrong If we go with the 2 radio button option, I would suggest that we need more "action" guidance in the field label, perhaps "Associate this cohort group with a content group:"  (which brings us right back to the original "Yes" and "No" options :-))
Can we by default have "No content group" selected so that people who aren't ready to make that associate don't have to do anything?

clrux commented on lms/static/js/groups/collections/cohort.js:
```diff
@@ -0,0 +1,16 @@
+var edx = edx || {};
+
+(function(Backbone, CohortModel) {
+    'use strict';
```
@cahrens To your thoughts:
* I think the text size should match the radio button label. I'll adjust.
* I've made adjustments to the styling of the visual feedback messages in a previous commit.
* I'll play with tab size to make them a tiny bit smaller, but I rather like larger click targets.

clrux replied:
@cahrens These should be addressed in commit: https://github.com/edx/edx-platform/commit/49c48b265561d86913b9ca45e55bc4927ff969a8

explorerleslie commented:
@catong Yeah, we could do something like that:

Associate this cohort group with a content group:
No content group
Select a content group: ______

explorerleslie commented:
@andy-armstrong @talbs @clrux (I don't remember who's UX point on the FED at this point) couple of minor pieces of feedback: 

* The green "Save" confirmation message looks smashed up against the tabs, which looks odd considering how much breathing room we give everything else in this area. Could we put some whitespace there? 
* I agree that the message "The selected content group has been deleted, you may wish to reassign this cohort group." is too small. We should also put some space after the icon and before the text. 
* For a deleted content group in the dropdown, we show "Some content group that's been deleted" @catong should provide a better string for that case.
* I think I mentioned this before, but I still want to gray out the dropdown for content group if the "No" option is selected. We should have an explicit action to switch from the No to Yes option.

clrux commented:
@andy-armstrong @talbs @explorerleslie I made a commit a bit ago that addresses some of the visual styling mentioned here.

catong commented:
@explorerleslie re the "Some content group that's been deleted" should already be addressed in some commit. There was some conversation about it with @andy-armstrong and @cahrens. We already use "Deleted content group" elsewhere and will go with that string here as well.

andy-armstrong commented:
@explorerleslie @catong I think your suggested wording is much better. My only question now is whether the label capitalization is inconsistent. There are now two fields:

    Cohort Group Name
    Associate this cohort group with a content group

If we're going more conversational, should I change the former? Or if not, I think we need something more like:

     Associated Content Group

catong commented:
@andy-armstrong I think you're right from a consistency perspective. @explorerleslie it's up to you. 
Also, we now call them "cohorted content groups" when we create them on the Group Configurations page. Do we live with the inconsistency of referring to them elsewhere by the shorter "content group" form? Agh....

andy-armstrong commented:
@explorerleslie @catong @cahrens @dan-f I've addressed all the feedback and updated my sandbox:

http://andy-armstrong.m.sandbox.edx.org/courses/AndyA/ACTC/1/instructor#view-membership

Please re-review when you have a chance.

@cahrens there are still Bok Choy failures that I'll investigate tomorrow if you don't beat me to it. I believe that all the Jasmine tests should be passing now.

andy-armstrong commented:
@clrux @talbs I forgot to tag you before... my sandbox has been updated:

http://andy-armstrong.m.sandbox.edx.org/courses/AndyA/ACTC/1/instructor#view-membership

explorerleslie commented:
I'm good with "Associated Cohorted Content Group" (I never remember if we title-case or sentence-case field labels), @catong @andy-armstrong

talbs commented on lms/templates/ux/reference/instructor_dashboard/cohorts.html:
```diff
@@ -0,0 +1,189 @@
+                          <label for="cohort-assn-course-yes"><input type="radio" name="cohort-assn-course-2" value="yes" checked="checked" /> Yes</label>
+                          <div class="input-group-other is-visible">
+                            <div class="msg error inline">
+                              <p class="copy-error"><i class="icon icon-warning-sign"></i> You haven't configured any content groups yet. You need to create a content group before you can create assignments. <a href="#">Create a content group</a></p>
```
Sweetness, thanks!

dan-f commented on common/test/acceptance/pages/lms/instructor_dashboard.py:
```diff
@@ -184,6 +194,68 @@ def get_cohort_student_input_field_value(self):
+        Returns all the content groups available for associating with the cohort currently being edited.
+        """
+        select = self.q(css=self._bounded_selector(self.content_group_selector))
+        groups = []
```
Could be a list comprehension:
```python
return [
    option if option.text != ""
    for option in self.q(css=self._bounded_selector(self.content_group_selector))
]
```

cahrens commented:
@explorerleslie and @catong what do you think of the select being empty before you have selected a content group? I would prefer there be some text in there, perhaps "No Content Group Selected".

![image](https://cloud.githubusercontent.com/assets/484484/5630911/0687244c-9590-11e4-8593-945a7e01a108.png)

dan-f commented on common/test/acceptance/tests/discussion/helpers.py:
```diff
@@ -68,11 +69,11 @@ def add_manual_cohort(self, course_fixture, cohort_name):
-        url = LMS_BASE_URL + "/courses/" + course_fixture._course_key + '/cohorts/add'
-        data = {"name": cohort_name}
+        url = LMS_BASE_URL + "/courses/" + course_fixture._course_key + '/cohorts/'
+        data = json.dumps({"name": cohort_name})
```
Why did you `json.dumps` the data? The django client expects a dict (https://docs.djangoproject.com/en/1.7/topics/testing/tools/#django.test.Client.post)

cahrens replied:
I'll try not doing json.dumps, but I don't think it worked without this. I'll let you know the details after I try it.

cahrens replied:
@dan-f My view method expects JSON, and I get an invalid JSON error if I don't convert the data to JSON. the Django client does not do the conversion for you.

cahrens commented:
@andy-armstrong The behavior is unexpected if you press "Save" with a deleted content group selected. I would have expected Save to be disabled (best) or at least an error message. Instead, it just selects another group and saves that.

dan-f commented on common/test/acceptance/tests/discussion/test_cohort_management.py:
```diff
@@ -82,18 +83,40 @@ def test_cohort_description(self):
+            self.cohort_management_page.get_cohort_related_content_group_message()
+        )
+        # TODO: test can't select radio button
+        # TODO: test link to Studio
```
What's the status on these TODO's? Looks like the second is taken care of in the below test case.

cahrens replied:
These are the bok choy tests that need to be finished, mentioned in standup. And the link to Studio mentioned here is different-- it takes you directly to the group configurations settings page, if no content groups have been defined.

andy-armstrong commented:
@cahrens Good catch. I thought I'd covered that case but I see I didn't write a test for it. I was thinking that the save would just work and it would continue to have the deleted content group selected. In general for editors I like them to be able to save a previously saved value even if it is an error, because you might be editing to change a different value. However, for now we don't support editing any other fields, so I'd be okay with disallowing the save. @explorerleslie do you have an opinion?

cahrens commented:
@clrux The error text isn't responsive (doesn't wrap).

![image](https://cloud.githubusercontent.com/assets/484484/5631151/37fdd4d8-9592-11e4-89c6-a2f18fed132e.png)

cahrens commented on common/test/acceptance/pages/lms/instructor_dashboard.py:
```diff
@@ -184,6 +194,68 @@ def get_cohort_student_input_field_value(self):
+        Returns an array of messages related to modifying cohort settings. If wait_for_messages
+        is True, will wait for a message to appear.
+        """
+        # Note that the class name should change because it is no longer a "create"
```
I should remove this comment.

cahrens commented on common/test/acceptance/pages/lms/instructor_dashboard.py:
```diff
@@ -184,6 +194,68 @@ def get_cohort_student_input_field_value(self):
+        Selects the specified content group from the selector. Assumes that content_group is not None.
+        """
+        option_selector = self.q(css=self._bounded_selector(self.content_group_selector))
+        option_selector.filter(lambda el: el.text == content_group).first.click()
```
One of the Solutions PRs I reviewed had a helper method for dealing with selects. I should see about bringing that into this PR.

cahrens commented on lms/static/js/groups/views/cohort_editor.js:
```diff
@@ -18,11 +27,38 @@
             this.$el.html(this.template({
                 cohort: this.model,
-                advanced_settings_url: this.advanced_settings_url
+                contentGroups: this.contentGroups,
```
Please remove contentGroups if not needed for the template.

cahrens commented on lms/static/js/groups/views/cohort_editor.js:
```diff
@@ -18,11 +27,38 @@
+            this.cohortFormView = new CohortFormView({
+                model: this.model,
+                contentGroups: this.contentGroups,
+                context: this.context
```
Just pass this.context.studioGroupConfigurationsUrl to the view (which is consistent with what you do on line 31)?

andy-armstrong replied:
What I decided to do was pass a context to each view rather than having to manually pass around all the different URLs that are needed. Even though this view currently only uses one item from the context, I prefer to be consistent about passing the whole context through.

Note that line 31 is different as that is a rendering of a template. I decided to not pass the context to the templates but to instead leave the URLs passed in directly. I'd be willing to change them to all receive a context instead if that seems more consistent.

cahrens replied:
Oh, I see. It's fine the way it is.

cahrens commented on lms/static/js/groups/views/cohort_editor.js:
```diff
@@ -18,11 +27,38 @@
+            var tabElement = $(event.currentTarget),
+                tabName = tabElement.data('tab');
+            event.preventDefault();
+            this.$('.wrapper-tabs .tab').removeClass('is-selected');
```
In some of the Solutions PRs, I've seen Aria roles being update as well. We should have a task for @cptvitamin to do an accessibility review. It might be good for him to take a look at your sandbox now.

dan-f commented on lms/static/js/spec/groups/views/cohorts_spec.js:
```diff
@@ -0,0 +1,657 @@
+
+            createMockCohort = function (name, id, userCount, groupId, userPartitionId) {
+                return {
+                    id: id || 1,
```
What about when `id` is 0?

andy-armstrong replied:
Good catch. I'll change it to test for null.

cahrens commented on lms/static/js/groups/views/cohort_form.js:
```diff
@@ -0,0 +1,138 @@
+                groupsEnabled = target.val() === 'yes';
+            if (!groupsEnabled) {
+                // If the user has chosen 'no', then clear the selection by setting
+                // it to the first option ('Choose a content group to associate').
```
Change the comment to reference the correct text (or something generic enough so that it doesn't depend on the text).

dan-f commented on lms/static/js/spec/groups/views/cohorts_spec.js:
```diff
@@ -0,0 +1,657 @@
+                return {
+                    id: id || 1,
+                    name: name,
+                    user_count: userCount || 0,
```
Why are we expecting that `userCount` could be undefined, but subsequent arguments are defined?

andy-armstrong replied:
Because the two subsequent arguments are both optional and so can be null.

explorerleslie commented:
@cahrens @talbs I actually prefer the selection to be blank, so as to not confuse with the "No content group" radio button above. However, we probably need an error state that says that you must select a content group.

clrux commented:
-- Please select --

On Tue, Jan 6, 2015 at 11:15 AM, Leslie Gerhat <notifications@github.com>
wrote:

> @cahrens <https://github.com/cahrens> @talbs <https://github.com/talbs> I
> actually prefer the selection to be blank, so as to not confuse with the
> "No content group" radio button above. However, we probably need an error
> state that says that you must select a content group.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/edx/edx-platform/pull/6308#issuecomment-68886856>.
>



-- 
Chris Rodriguez | Front End Developer
141 Portland Street, 9th Floor
Cambridge, MA 02139
607-333-1667
crodriguez@edx.org
@clrux

cahrens commented on lms/static/js/spec/groups/views/cohorts_spec.js:
```diff
@@ -0,0 +1,657 @@
+                        expect(cohortsView.$('.radio-yes').prop('disabled')).toBeTruthy();
+                        expect(
+                            cohortsView.$('.msg-inline').text().trim(),
+                            'No content groups exist. Create a content group to associate with cohort groups.'
```
Test the link too?

cahrens commented on lms/static/js/spec/groups/views/cohorts_spec.js:
```diff
@@ -0,0 +1,657 @@
+
+            describe("Cohort Settings", function() {
+                describe("Content Group Setting", function() {
+                    it("shows a select element with an option for each content group", function () {
```
Test specifying a content group at time of creation (as opposed to going to settings)?

andy-armstrong replied:
I do this in the separate add tests: https://github.com/edx/edx-platform/blob/christina/link_group_content_to_cohort/lms/static/js/spec/groups/views/cohorts_spec.js#L265

dan-f commented on lms/static/js/spec/groups/views/cohorts_spec.js:
```diff
@@ -0,0 +1,657 @@
+                TemplateHelpers.installTemplate('templates/file-upload');
+            });
+
+            it("Show an error if no cohorts are defined", function() {
```
Nit: Show -> shows (also, don't we normally use lower-case for individual test cases?)

andy-armstrong replied:
Yes, we do. The intention is that it reads with the 'it', e.g. 'it shows an error if no cohorts are defined'. I've fixed it.

dan-f commented on lms/static/js/spec/groups/views/cohorts_spec.js:
```diff
@@ -0,0 +1,657 @@
+                expect(cohortsView.$('.wrapper-cohort-supplemental')).toHaveClass('is-hidden');
+            });
+
+            it("Syncs data when membership tab is clicked", function() {
```
Similar capitalization nit.

catong commented on lms/templates/instructor/instructor_dashboard_2/cohort-editor.underscore:
```diff
@@ -1,63 +1,72 @@
+            <div class="form-fields">
+                <div class="field field-textarea is-required">
+                    <label for="cohort-management-group-add-students" class="label">
+                        <%- gettext('Enter email addresses and/or usernames separated by new lines or commas for students to add. *') %>
```
"Enter email addresses and/or usernames, separated by new lines or commas, for the students you want to add."

dan-f commented on lms/static/js/spec/groups/views/cohorts_spec.js:
```diff
@@ -0,0 +1,657 @@
+                verifyHeader(1, 'Cat Lovers', catLoversInitialCount);
+                $(cohortsView.getSectionCss("membership")).click();
+                AjaxHelpers.expectRequest(requests, 'GET', '/mock_service/cohorts');
+                respondToRefresh(1001, 2);
```
I found this a little hard to read. Maybe rename it to to more clearly reflect that it's returning cohort json, and that the arguments indicate cat count and dog count respectively? Maybe something like `respondWithCohortJson({catCount: 1001, dogCount: 2})`

andy-armstrong replied:
This is an old test from an earlier PR. It only looks like I added it because I moved the file. It looks like somehow my refactoring lost the history of the file though... I'll see if I can fix that up.

cahrens commented:
@andy-armstrong a few minor comments, but otherwise :+1: (I don't need to re-review).

I'll work on cleaning up my stuff now and finishing that one bok choy test.

talbs commented:
@clrux, a few visual tweaks/feedback notes (in order of relative importance to fix):

**The font-size for the inline message and the radio button labels seems a bit big (when compared to the "Associate this cohort group with a content group" header). Please make the equivalent of 14px**
![screen shot 2015-01-06 at 12 22 54 pm](https://cloud.githubusercontent.com/assets/163763/5632719/ee5fc216-959e-11e4-9400-81881f6c73b8.png)

**The border-width on the active tab and bottom of the tab nav is too thick - please reduce.**
![screen shot 2015-01-06 at 12 18 12 pm](https://cloud.githubusercontent.com/assets/163763/5632713/d3b1c1b2-959e-11e4-93c7-8e6016145166.png)

**The visual horizontal margin on this error message seems off. It should span the width of its parent**
![screen shot 2015-01-06 at 12 18 45 pm](https://cloud.githubusercontent.com/assets/163763/5632867/39eb57a8-95a0-11e4-9f65-4b41b1bb8c39.png)

Mind addressing these?

talbs commented on lms/templates/instructor/instructor_dashboard_2/cohort-form.underscore:
```diff
@@ -0,0 +1,99 @@
+            %>
+            <div class="form-field">
+                <div class="cohort-management-details-association-course field field-radio">
+                    <label class="label">
```
@clrux, this is an improper use of the ``<label>`` element. It should have a ``for="foo"`` attribute and relate to a particular element's ``id`` attribute. If this is more of a heading, I'd recommend you change it to use a heading element (with an appropriate heading-related semantic class, i.e. not "label") and just migrate any Sass/styling to that.

cptvitamin replied:
:+1:

clrux replied:
Good catch, not sure why that was there. It's a heading now. However, it looks like the class `.label` is pre-existing and used in `.member-list-widget`. Not sure how pervasive this is, but I've added an additional class to the styles for this particular case.

talbs commented on lms/templates/instructor/instructor_dashboard_2/cohorts.underscore:
```diff
@@ -24,6 +24,9 @@
 </div>
 
+<!-- Add modal -->
+<div class="cohort-management-add-modal"></div>
```
NIT: I'd sway away from using the term modal in class names and in describing this element. From a UX/UI PoV, modal implies taking focus and constraining it to a particular DOM element (hence the usual overlay + save/cancel actions). Any chance we could use a term more suitable to this case, like ``cohort-management-add-form`` or ``cohort-management-add``?

talbs commented on lms/static/sass/course/instructor/_cohort-management.scss:
```diff
@@ -0,0 +1,143 @@
+        overflow: auto;
+        width: 100%;
+        height: auto;
+        margin: 0 0 0 $baseline;
```
Add [RTL](https://github.com/edx/edx-platform/wiki/RTL-UI-Best-Practices) support.

clrux replied:
This file is obsolete. Removed in upcoming commit.

talbs commented on lms/static/sass/course/instructor/_cohort-management.scss:
```diff
@@ -0,0 +1,143 @@
+          width: flex-grid(3);
+          vertical-align: top;
+
+          .float-right {
```
I thought I made a previous comment about this rule. Why can't this property be placed on a specific UI element's class name rather than be a utility class. This seems like a bad idea to 1) creating utility styling for basic floating at this point, and 2) add into this layer of our Sass (scoped to a particular view). Please remove and refactor.

clrux replied:
This file is obsolete. Removed in upcoming commit.

talbs commented on lms/static/sass/course/instructor/_cohort-management.scss:
```diff
@@ -0,0 +1,143 @@
+          vertical-align: top;
+
+          .float-right {
+            float: right;
```
NIT: when you do refactor this property, please add [RTL](https://github.com/edx/edx-platform/wiki/RTL-UI-Best-Practices) support.

clrux replied:
This file is obsolete. Removed in upcoming commit.

talbs commented on lms/static/sass/course/instructor/_cohort-management.scss:
```diff
@@ -0,0 +1,143 @@
+
+      .cohort-section-header {
+        margin-top: ($baseline*1.5);
+        padding: $baseline 0 ($baseline/2) 0;
```
Do you need to reset horizontal padding to 0 here or can you use ``padding-top`` and ``padding-bottom`` rules?

clrux replied:
This file is obsolete. Removed in upcoming commit.

talbs commented on lms/static/sass/course/instructor/_instructor_2.scss:
```diff
@@ -53,6 +52,19 @@
+  .msg-inline {
+    // @extend %t-copy-sub1;
+    display: inline-block;
+    margin: 0 0 0 $baseline;
```
Add [RTL](https://github.com/edx/edx-platform/wiki/RTL-UI-Best-Practices) support.

talbs replied:
Do you need to reset the other three margin properties here or would ``margin-left`` be better suited?

talbs commented on lms/static/sass/course/instructor/_instructor_2.scss:
```diff
@@ -53,6 +52,19 @@
+
+    .icon,
+    .fa {
+      margin-right: ($baseline/2);
```
Add [RTL](https://github.com/edx/edx-platform/wiki/RTL-UI-Best-Practices) support.

talbs commented on lms/static/sass/course/instructor/_instructor_2.scss:
```diff
@@ -537,11 +545,17 @@ section.instructor-dashboard-content-2 {
     margin-bottom: $baseline;
 
+    .message {
+      margin-left: $baseline;
```
Add [RTL](https://github.com/edx/edx-platform/wiki/RTL-UI-Best-Practices) support.

talbs commented on lms/static/sass/course/instructor/_instructor_2.scss:
```diff
@@ -537,11 +545,17 @@ section.instructor-dashboard-content-2 {
 
+    .message {
+      margin-left: $baseline;
+      margin-right: $baseline;
```
Add [RTL](https://github.com/edx/edx-platform/wiki/RTL-UI-Best-Practices) support.

talbs commented on lms/static/sass/course/instructor/_instructor_2.scss:
```diff
@@ -550,7 +564,7 @@ section.instructor-dashboard-content-2 {
     }
 
-    .form-fields {
+    .form-field {
```
Why was this changed?

clrux replied:
We were trying to be a little more semantic, as this an individual field within the parent `.form-fields`.

talbs commented on lms/static/sass/course/instructor/_instructor_2.scss:
```diff
@@ -956,7 +973,148 @@ section.instructor-dashboard-content-2 {
     }
   }
+
+  /*
```
@clrux, where do we stand with this override being needed/still relevant. I'd like to be able to view the actual Sass you are planning on introducing into the final production code. Does that mean this needs to be refactored/removed?

talbs replied:
FWIW, I'm going to review this code to be thorough. Please take my comments in mind as you refactor.

clrux replied:
Sorry, this was already merged from _cohorts-management.scss, I just neglected to remove the comment.

talbs commented on lms/static/sass/course/instructor/_instructor_2.scss:
```diff
@@ -956,7 +973,148 @@ section.instructor-dashboard-content-2 {
+      overflow: auto;
+      width: 100%;
+      height: auto;
+      margin: 0 0 0 $baseline;
```
Add [RTL](https://github.com/edx/edx-platform/wiki/RTL-UI-Best-Practices) support.

talbs commented on lms/static/sass/course/instructor/_instructor_2.scss:
```diff
@@ -956,7 +973,148 @@ section.instructor-dashboard-content-2 {
+      border-bottom: 0;
+
+      .group-header-title {
+        padding: $baseline;
```
NIT: property order

talbs commented on lms/static/sass/course/instructor/_instructor_2.scss:
```diff
@@ -956,7 +973,148 @@ section.instructor-dashboard-content-2 {
+    background: $gray-l5;
+
+    .cohort-management-group-header {
+      padding: 0;
```
NIT: property order

talbs commented on lms/static/sass/course/instructor/_instructor_2.scss:
```diff
@@ -956,7 +973,148 @@ section.instructor-dashboard-content-2 {
+          display: inline-block;
+          width: flex-grid(3);
+          vertical-align: top;
+          text-align: right;
```
Add [RTL](https://github.com/edx/edx-platform/wiki/RTL-UI-Best-Practices) support.

talbs commented on lms/static/sass/course/instructor/_instructor_2.scss:
```diff
@@ -956,7 +973,148 @@ section.instructor-dashboard-content-2 {
+      a {
+        display: inline-block;
+        padding: ($baseline - 5);
+        -webkit-transition: none;
```
Please use the [Bourbon transition mixin](http://bourbon.io/docs/#transitions) here.

clrux replied:
Even if zeroing out already set transitions?

talbs commented on lms/static/sass/course/instructor/_instructor_2.scss:
```diff
@@ -956,7 +973,148 @@ section.instructor-dashboard-content-2 {
+    @extend %ui-no-list;
+    @extend %ui-depth1;
+    position: relative;
+    top: (($baseline/5));
```
double-bagging your Sass variable here, guy. No need to do that :P

talbs commented on lms/static/sass/course/instructor/_instructor_2.scss:
```diff
@@ -956,7 +973,148 @@ section.instructor-dashboard-content-2 {
 }
+/*
+ * End of additions/overrides
+ * Don't forget to refactor.
```
Never forget.

dan-f commented on openedx/core/djangoapps/course_groups/tests/test_views.py:
```diff
@@ -195,94 +225,197 @@ def test_default_cohort(self):
+        )
+        self.assertEqual((None, None), get_group_info_for_cohort(self.cohort1))
+        self.assertEqual(None, response_dict.get("group_id"))
+        self.assertEqual(None, response_dict.get("user_partition_id"))
```
Can use `self.assertIsNone` for the above two lines.

dan-f commented:
I'm done with my second pass. Everything looks fantastic! Great tests!

explorerleslie commented:
@clrux oh, yes, we can go with "-- Please select --" as the empty dropdown item, as long as you can't save that option as a selection. @andy-armstrong

andy-armstrong commented:
I had a chat with @talbs about the labeling and behavior of the select. Here are his recommendations:

1. Use "Not selected" as the label (see http://ux.stackexchange.com/questions/38928/adding-select-as-first-option-placeholder-of-dropdown-list)
2. Don't allow "Not selected" to be selected (http://stackoverflow.com/questions/8605516/default-select-option-as-blank)
3. Show an error if a value isn't selected, or if the value is a deleted cohorted content group.

Thoughts, @explorerleslie @clrux @cahrens?

talbs commented:
@clrux, I've completed a second pass. Let me know when you want me to confirm changes and sign off.

explorerleslie commented:
I like that approach for the behavior of the select, @andy-armstrong @talbs

catong commented:
@explorerleslie @andy-armstrong @talbs @clrux if I might chime in: just want to strongly suggest that we don't use "please" in the UI (@andy-armstrong has heard my thoughts about this!)

talbs commented:
@catong, so what you're saying is "Please, do not use 'please' in our application's UI?" :P

I'm fine with that tone/copy rule. Thanks much for the help.

catong commented:
@talbs yes, please!

cptvitamin commented on lms/templates/ux/reference/instructor_dashboard/cohorts.html:
```diff
@@ -0,0 +1,189 @@
+
+                <div class="cohort-management-details">
+                  <form action="" method="post" name="" id="cohort-management-details-form" class="cohort-management-details-form">
+                    <h3 class="form-title sr">Cohort Group Details</h3>
```
I recommend that we stop using hidden (screen reader only/sr) headings.  I consider this to be a hack.  We should use aria-labels and aria-labelled by when a label is actually needed.  In this case, no label is required.

cptvitamin commented on lms/templates/ux/reference/instructor_dashboard/cohorts.html:
```diff
@@ -0,0 +1,188 @@
+                      <div class="cohort-management-details-name field field-text">
+                        <h4 class="cohort-section-header no-top-margin">Cohort Group Name:</h4>
+                        <label for="cohort-details-name" class="label sr">Cohort Group Name</label>
+                        <input type="text" name="cohort-details-name" value="Cohort Name is Placed Here and Should Accommodate Almost Everything" class="input cohort-details-name" id="cohort-details-name" placeholder="Enter Your Cohort's Name" required />
```
A screen reader user hears "Cohort Group Name" twice here.  I recommend getting rid of the ```<h4>``` and styling the ```<label>``` to look like the ```<h4>``` did.  If you want to get rid of the hassle of managing the for/ID relationship, just wrap the ```<label>``` around the input.

cptvitamin commented on lms/templates/ux/reference/instructor_dashboard/cohorts.html:
```diff
@@ -0,0 +1,188 @@
+                    <div class="form-fields">
+                      <div class="cohort-management-details-assignment field field-radio">
+                        <label for="cohort-assign-random"><input type="radio" name="cohort-assignment" value="random" /> Randomly Assigned</label>
+                        <label for="cohort-assign-manual"><input type="radio" name="cohort-assignment" value="menual" checked="checked" /> Manually Assigned</label>
```
typo in ```value``` value

clrux replied:
Heh. Menual.

cptvitamin commented on lms/templates/ux/reference/instructor_dashboard/cohorts.html:
```diff
@@ -0,0 +1,188 @@
+                      </div>
+                    </div>
+
+                    <h4 class="cohort-section-header divider-lv2">Students in this cohort group are:</h4>
```
To me, this calls for a ```<fieldset>/<legend>``` rather than an ```<h4>```  Navigation aids like Headings are not designed to be used in forms unless it is a *very* large form or a form in which only a portion of it is expected to be filled out.   A screen reader user enters "form processing mode" and doesn't even encounter Headings, which is why its important to use standard form controls like ```labels``` and ```fieldsets``` to identify and demarcate form controls.

cptvitamin commented on lms/templates/ux/reference/instructor_dashboard/cohorts.html:
```diff
@@ -0,0 +1,188 @@
+                    <h4 class="cohort-section-header divider-lv2">Is this cohort group associated with course-based content groups?</h4>
+                    <div class="form-fields">
+                      <div class="cohort-management-details-assn-course field field-radio">
+                        <label for="cohort-assn-course-no"><input type="radio" name="cohort-assn-course" value="no" /> No</label>
```
no need for ```for``` attribute.  It's implied and currently non-conformant (takes value of ID, not class)

cptvitamin commented on lms/templates/ux/reference/instructor_dashboard/cohorts.html:
```diff
@@ -0,0 +1,188 @@
+                      <div class="cohort-management-details-assn-course field field-radio">
+                        <label for="cohort-assn-course-no"><input type="radio" name="cohort-assn-course" value="no" /> No</label>
+                        <div class="input-group has-other">
+                          <label for="cohort-assn-course-yes"><input type="radio" name="cohort-assn-course" value="yes" checked="checked" /> Yes</label>
```
see above

cptvitamin commented on lms/templates/instructor/instructor_dashboard_2/cohort-editor.underscore:
```diff
@@ -1,63 +1,72 @@
-</div>
+            <div class="form-actions">
+                <button class="form-submit button action-primary action-view">
+                    <i class="button-icon icon icon-plus"></i> <%- gettext('Add Students') %>
```
When accessible text of equal value is adjacent to the icon (99% of cases here at edX) we should add ```aria-hidden="true"``` to the icon fonts.  Some screen readers (Like Voiceover) will focus on them and remain silent, which makes screen reader users wonder if they are missing important content.  It also visually renders as a generic glyph in screen readers that are also used by people with cognitive disabilities (like Voiceover).  This should do the trick:
```<i class="button-icon icon icon-plus" aria-hidden="true"></i>```  <--this is one of the very few exceptions for using aria-hidden at all
@clrux add this note to your 'bag-o-patterns' when you get around to it

cptvitamin commented on lms/templates/ux/reference/instructor_dashboard/cohorts.html:
```diff
@@ -0,0 +1,188 @@
+                    <!-- // -->
+
+                    <!-- When no options are available. -->
+                    <h4 class="cohort-section-header divider-lv2">Is this cohort group associated with course-based content groups?</h4>
```
```<fieldset>/<legend>``` see above comment

cptvitamin commented on lms/templates/ux/reference/instructor_dashboard/cohorts.html:
```diff
@@ -0,0 +1,188 @@
+                  <div class="cohort-management-file-upload csv-upload is-hidden" id="cohort-management-file-upload"></div>
+
+                  <div class="cohort-management-supplemental">
+                      <p class=""><i class="icon icon-info-sign"></i> To review student cohort group assignments or see the results of uploading a CSV file, download course profile information or cohort results on <a href="" class="link-cross-reference" data-section="data_download"> the Data Download page</a>.</p>
```
This is an example of when text that represents the icon is not adjacent. Here, I recommend adding the ```aria-hidden="true"``` to the icon and then adding screen reader only text "Note:"

clrux commented:
@talbs I've gone through and addressed most, if not all of your feedback. Let's discuss the use of `fieldset` at some point.
@cptvitamin Thanks for the a11y heads-up. I've addressed `icon` hidden items and will discuss the `fieldset` stuff with the team.

biomassives commented:
This message was created automatically by mail delivery software.

A message that you sent could not be delivered to one or more of its
recipients. This is a temporary error. The following address(es) deferred:

  acmeideal@gmail.com
    Domain biomassiv.es has exceeded the max emails per hour (104/100 (104%)) allowed.  Message will be reattempted later

------- This is a copy of the message, including all the headers. ------
Received: from o4.sgmail.github.com ([192.254.112.99]:5612)
	by server23.websitehostserver.net with esmtps (UNKNOWN:DHE-RSA-AES128-GCM-SHA256:128)
	(Exim 4.82)
	(envelope-from <bounces+848413-b37a-greg=biomassiv.es@sgmail.github.com>)
	id 1Y8c9X-000GUZ-8B
	for greg@biomassiv.es; Tue, 06 Jan 2015 15:59:04 -0600
DKIM-Signature: v=1; a=rsa-sha1; c=relaxed; d=github.com; 
	h=from:reply-to:to:in-reply-to:references:subject:mime-version:content-type:content-transfer-encoding:list-id:list-archive:list-post:list-unsubscribe; 
	s=smtpapi; bh=qT+c7HVyV5CGqKA4f1NsgfNa05U=; b=a6NifSE7kaS1i24A2u
	0W+oz6DVWgLgIPJTG0//kCpSjHLfQx4iL5yazhyUlMG3XoVk8qSlGIoex//vfadX
	hbNm3XZprLS9X/UWm68kU3MotGxEx4tMF1wVLTv4GPog8NPjfofa3zU19sKkbx2F
	b3E/0ahgTlu9fDHVzQQ/UrmEQ=
Received: by filter-379.sjc1.sendgrid.net with SMTP id filter-379.25139.54AC5AA62F
        2015-01-06 21:59:03.323695006 +0000 UTC
Received: from github-smtp2b-ext-cp1-prd.iad.github.net (github-smtp2b-ext-cp1-prd.iad.github.net [192.30.253.17])
	by ismtpd-045 (SG) with ESMTP id 14ac128aafa.5730.3b44d
	for <greg@biomassiv.es>; Tue, 06 Jan 2015 21:31:15 +0000 (UTC)
Date: Tue, 06 Jan 2015 13:31:14 -0800
From: Chris <notifications@github.com>
Reply-To: edx/edx-platform <reply@reply.github.com>
To: edx/edx-platform <edx-platform@noreply.github.com>
Message-ID: <edx/edx-platform/pull/6308/c68937381@github.com>
In-Reply-To: <edx/edx-platform/pull/6308@github.com>
References: <edx/edx-platform/pull/6308@github.com>
Subject: Re: [edx-platform] Assign cohorts to groups in the instructor
 dashboard (#6308)
Mime-Version: 1.0
Content-Type: multipart/alternative;
 boundary="--==_mimepart_54ac5422cecf8_4c7b3fb2a0b392b8132518f";
 charset=UTF-8
Content-Transfer-Encoding: 7bit
Precedence: list
X-GitHub-Sender: clrux
X-GitHub-Recipient: biomassives
List-ID: edx/edx-platform <edx-platform.edx.github.com>
List-Archive: https://github.com/edx/edx-platform
List-Post: <mailto:reply@reply.github.com>
List-Unsubscribe: <mailto:unsub+0042d4e2cc039e26ff992e5b3b2c8a8ef04fbaffc8bbf18692cf0000000110c4162292a169ce031f8c29@reply.github.com>,
 <https://github.com/notifications/unsubscribe/AELU4oPwVKUBjKtsoO7g_kBtAojASu6Fks5nfEuigaJpZM4DKLhP>
X-Auto-Response-Suppress: All
X-GitHub-Recipient-Address: greg@biomassiv.es
X-SG-EID: ZlaDthyue/4HG+EgTb7JoZYs48JNb2iQtCcPV6oIs9X4776/JHRCD+BsXd2BuXdPHbkgKpVezulKgW
 89qUBcBCyGbbtjuT633FNBbHbPS6QddUSZGBdB8l4tUu1OnPxfMBvKOkrwmGMQteCr/osIgN5IF8/s
 /fUIZfi+w3ezJ/0=
X-Spam-Status: No, score=0.1
X-Spam-Score: 1
X-Spam-Bar: /
X-Ham-Report: Spam detection software, running on the system "server23.websitehostserver.net", has
 identified this incoming email as possible spam.  The original message
 has been attached to this so you can view it (if it isn't spam) or label
 similar future email.  If you have any questions, see
 root\@localhost for details.
 
 Content preview:  @talbs I've gone through and addressed most, if not all of
    your feedback. Let's discuss the use of `fieldset` at some point. @cptvitamin
    Thanks for the a11y heads-up. I've addressed `icon` hidden items and will
    discuss the `fieldset` stuff with the team. [...] 
 
 Content analysis details:   (0.1 points, 5.0 required)
 
  pts rule name              description
 ---- ---------------------- --------------------------------------------------
 -0.0 T_RP_MATCHES_RCVD      Envelope sender domain matches handover relay
                             domain
  0.0 URIBL_BLOCKED          ADMINISTRATOR NOTICE: The query to URIBL was blocked.
                             See
                             http://wiki.apache.org/spamassassin/DnsBlocklists#dnsbl-block
                              for more information.
                             [URIs: github.com]
 -1.9 BAYES_00               BODY: Bayes spam probability is 0 to 1%
                             [score: 0.0000]
  2.1 HTML_IMAGE_ONLY_12     BODY: HTML: images with 800-1200 bytes of words
  0.0 HTML_MESSAGE           BODY: HTML included in message
 -0.1 DKIM_VALID_AU          Message has a valid DKIM or DK signature from author's
                             domain
  0.1 DKIM_SIGNED            Message has a DKIM or DK signature, not necessarily valid
 -0.1 DKIM_VALID             Message has at least one valid DKIM or DK signature
X-Spam-Flag: NO

----==_mimepart_54ac5422cecf8_4c7b3fb2a0b392b8132518f
Content-Type: text/plain;
 charset=UTF-8
Content-Transfer-Encoding: 7bit

@talbs I've gone through and addressed most, if not all of your feedback. Let's discuss the use of `fieldset` at some point.
@cptvitamin Thanks for the a11y heads-up. I've addressed `icon` hidden items and will discuss the `fieldset` stuff with the team.

---
Reply to this email directly or view it on GitHub:
https://github.com/edx/edx-platform/pull/6308#issuecomment-68937381
----==_mimepart_54ac5422cecf8_4c7b3fb2a0b392b8132518f
Content-Type: text/html;
 charset=UTF-8
Content-Transfer-Encoding: 7bit

<p><a href="https://github.com/talbs" class="user-mention">@talbs</a> I've gone through and addressed most, if not all of your feedback. Let's discuss the use of <code>fieldset</code> at some point.<br>
<a href="https://github.com/cptvitamin" class="user-mention">@cptvitamin</a> Thanks for the a11y heads-up. I've addressed <code>icon</code> hidden items and will discuss the <code>fieldset</code> stuff with the team.</p>

<p style="font-size:small;-webkit-text-size-adjust:none;color:#666;">&mdash;<br>Reply to this email directly or <a href="https://github.com/edx/edx-platform/pull/6308#issuecomment-68937381">view it on GitHub</a>.<img alt="" height="1" src="https://github.com/notifications/beacon/AELU4oo5VeMG2qSjbhahzmN0x3JoWpwlks5nfEuigaJpZM4DKLhP.gif" width="1" /></p>
<div itemscope itemtype="http://schema.org/EmailMessage">
  <div itemprop="action" itemscope itemtype="http://schema.org/ViewAction">
    <link itemprop="url" href="https://github.com/edx/edx-platform/pull/6308#issuecomment-68937381"></link>
    <meta itemprop="name" content="View Pull Request"></meta>
  </div>
  <meta itemprop="description" content="View this Pull Request on GitHub"></meta>
</div>

----==_mimepart_54ac5422cecf8_4c7b3fb2a0b392b8132518f--

andy-armstrong commented:
@explorerleslie @catong @cahrens @dan-f @talbs @clrux I've implemented validation of cohort settings in the instructor dashboard, and updated the sandbox:

http://andy-armstrong.m.sandbox.edx.org/courses/AndyA/ATCC/1/instructor#view-membership

I've also implemented Jasmine tests for all the cases discussed in this thread. @cahrens please re-review just my final commit. I also imagine that I might have broken some Bok Choy tests as I'm now using the detailed notification scheme.

talbs commented:
@clrux, to [your question above](https://github.com/edx/edx-platform/pull/6308#issuecomment-69022744), yes - please add the standard amount of padding from elsewhere.

clrux commented:
@talbs @cahrens Re: the responsiveness (or lack thereof) with the error message... the message itself should wrap - the container/grid isn't flexing. Has it always been this way?

cahrens commented:
@andy-armstrong a couple small comments in https://github.com/edx/edx-platform/commit/fcb0ed12ff0906c1d9fe691772c4118443c0726f, but looks great!

talbs commented:
@clrux, thanks for the revisions. Aside from the a11y fixes noted by @cptvitamin (which will be tackled in a later set of work), things look good.

:+1:
